base/reduce.jl

```diff
@@ -1379,3 +1328,65 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
     end
     return n
 end
+
+function _mapped_eltype(f, A)
+    elT = eltype(A) # try this
+    if elT === Any
+        elT = @default_eltype(A) # try again
+    end
+    fT = _return_type(f, Tuple{elT}) # determine type of f(A[i]), if possible
+    return isconcretetype(fT) ? fT : Any # always widen this to `Any` if non-concrete
```

Comment:
Why?

Reply:
Just a strong hint that this function is not that reliable.
As the current optimization only works for concrete cases.